risc0/bigint2/src/ec/mod.rs

```diff
@@ -160,31 +179,71 @@ impl<const WIDTH: usize, C: Curve<WIDTH>> AffinePoint<WIDTH, C> {
         }
 
         // Return the result, based on which buffer was written to last.
-        let result_point = if result_flip { result2 } else { result1 };
-        result.buffer = result_point;
+        *result = if result_flip { result2 } else { result1 };
     }
 
+    /// Elliptic curve doubling of the affine point.
     #[stability::unstable]
     pub fn double(&self, result: &mut Self) {
         let curve = C::CURVE;
-        double_raw(self.as_u32s(), curve.as_u32s(), &mut result.buffer);
+        // If the point is zero, can short-circuit and return the identity point.
+        if self.is_zero {
+            *result = *self;
+        } else {
+            if self.buffer[1] != [0u32; WIDTH] {
+                unsafe {
+                    double_raw(self.as_u32s(), curve.as_u32s(), &mut result.buffer);
+                }
+                // DO NOT REMOVE: the result is unchecked, and only the buffer is updated above
+                result.is_zero = false;
+            } else {
+                // DO NOT REMOVE: the result is unchecked, and only the buffer is updated above
```

Comment:
Not sure this comment is quite accurate. Maybe this?
```suggestion
                // DO NOT REMOVE: in this case a zero has been computed and the buffer is ignored
```